### PR TITLE
add go 1.22 build

### PR DESCRIPTION
**Pull request by lizzzcai**
add go 1.22 build support to fix some vulnerabilities.

### CODE REVIEW

File: e2e/run_test.go
Comment (igungor) [inline e2e/run_test.go:266]: Could you revert this change please? We'd like to fix test timeout issues in another PR where we set the `cmd.Timeout` in the `s5cmd` helper function.

File: storage/s3.go
Comment (igungor) [inline storage/s3.go:1052]: Why is this change necessary? Do you use a lint tool other than the `make check` tooling?

Comment (lizzzcai) [reply storage/s3.go:1052]: Hi @igungor , I get this error when trying `make check`, 

```
❯ go version
go version go1.22.6 darwin/arm64
❯ staticcheck -version
staticcheck 2024.1.1 (0.5.1)
❯ make check
storage/s3.go:1052:9: printf-style function with dynamic format string and no further arguments should use print-style function instead (SA1006)
make: *** [staticcheck] Error 1
```

File: e2e/run_test.go
Comment (lizzzcai) [reply e2e/run_test.go:266]: sure, I will revert it and let me know once your PR is merged, I will rebase it.

**Comment by igungor** [comment]
Now that we have Go 1.23.2, could you update your PR to
- use go 1.23.x in goreleaser and docker builds
- use 1.23.x, 1.22.x, and 1.21.x in github actions job matrices please? :)

File: storage/s3.go
Comment (igungor) [reply storage/s3.go:1052]: 👍

**Comment by lizzzcai** [comment]
> Now that we have Go 1.23.2, could you update your PR to
> 
> * use go 1.23.x in goreleaser and docker builds
> * use 1.23.x, 1.22.x, and 1.21.x in github actions job matrices please? :)

Done. I assumed 1.19 and 1.20 are not needed anymore.

**Comment by igungor** [comment]
> > Now that we have Go 1.23.2, could you update your PR to
> > 
> > * use go 1.23.x in goreleaser and docker builds
> > * use 1.23.x, 1.22.x, and 1.21.x in github actions job matrices please? :)
> 
> Done. I assumed 1.19 and 1.20 are not needed anymore.

Correct. We test test latest 3 Go releases.

**Comment by igungor** [comment]
`ci / qa (1.23.x, ubuntu)` fails because staticcheck tool doesn't recognize a Go 1.23 feature yet. Updating `	honnef.co/go/tools` to `v0.6.0-0.dev` fixes the issue.

I'd push the necessary changes but I don't have access to your repo.

Could you do the above version change in `go.mod` fileand run `go mod tidy && go mod vendor` please? This will bring the necessary changes. Please commit those changes and let's see if it's working.

**Comment by lizzzcai** [comment]
> `ci / qa (1.23.x, ubuntu)` fails because staticcheck tool doesn't recognize a Go 1.23 feature yet. Updating ` honnef.co/go/tools` to `v0.6.0-0.dev` fixes the issue.
> 
> I'd push the necessary changes but I don't have access to your repo.
> 
> Could you do the above version change in `go.mod` fileand run `go mod tidy && go mod vendor` please? This will bring the necessary changes. Please commit those changes and let's see if it's working.

Hi @igungor, by upgrading to `honnef. com/go/tools v0.6.0-0.dev`, the minimal Go version becomes `1.22.1`.

**Comment by igungor** [comment]
> > `ci / qa (1.23.x, ubuntu)` fails because staticcheck tool doesn't recognize a Go 1.23 feature yet. Updating ` honnef.co/go/tools` to `v0.6.0-0.dev` fixes the issue.
> > I'd push the necessary changes but I don't have access to your repo.
> > Could you do the above version change in `go.mod` fileand run `go mod tidy && go mod vendor` please? This will bring the necessary changes. Please commit those changes and let's see if it's working.
> 
> Hi @igungor, by upgrading to `honnef. com/go/tools v0.6.0-0.dev`, the minimal Go version becomes `1.22.1`.

Hmm, it may not be the time to add Go 1.23 to the build matrix :(

**Comment by lizzzcai** [comment]
> > > `ci / qa (1.23.x, ubuntu)` fails because staticcheck tool doesn't recognize a Go 1.23 feature yet. Updating ` honnef.co/go/tools` to `v0.6.0-0.dev` fixes the issue.
> > > I'd push the necessary changes but I don't have access to your repo.
> > > Could you do the above version change in `go.mod` fileand run `go mod tidy && go mod vendor` please? This will bring the necessary changes. Please commit those changes and let's see if it's working.
> > 
> > 
> > Hi @igungor, by upgrading to `honnef. com/go/tools v0.6.0-0.dev`, the minimal Go version becomes `1.22.1`.
> 
> Hmm, it may not be the time to add Go 1.23 to the build matrix :(

Hi @igungor , how about keep it to `1.22.x, 1.21.x and 1.20.x`? Overall 1.22.x meets my requirements.

**Comment by igungor** [comment]
> > > > `ci / qa (1.23.x, ubuntu)` fails because staticcheck tool doesn't recognize a Go 1.23 feature yet. Updating ` honnef.co/go/tools` to `v0.6.0-0.dev` fixes the issue.
> > > > I'd push the necessary changes but I don't have access to your repo.
> > > > Could you do the above version change in `go.mod` fileand run `go mod tidy && go mod vendor` please? This will bring the necessary changes. Please commit those changes and let's see if it's working.
> > > 
> > > 
> > > Hi @igungor, by upgrading to `honnef. com/go/tools v0.6.0-0.dev`, the minimal Go version becomes `1.22.1`.
> > 
> > 
> > Hmm, it may not be the time to add Go 1.23 to the build matrix :(
> 
> Hi @igungor , how about keep it to `1.22.x, 1.21.x and 1.20.x`? Overall 1.22.x meets my requirements.

sounds good.

**Comment by lizzzcai** [comment]
Hi @igungor , updated. any timeline for fixing the timeout issue in new go version? and is there any plan to trigger a new release in the main branch (I saw a similar question #766)? Thanks.

**Comment by igungor** [comment]
Thanks. Increased s5cmd test command timeout to 10 minutes but the issue is still there. I'm not sure what the problem is on Go 1.21 macos test.

**Comment by lizzzcai** [comment]
> Thanks. Increased s5cmd test command timeout to 10 minutes but the issue is still there. I'm not sure what the problem is on Go 1.21 macos test.

Hi @igungor , if I am not wrong, I think this line [overwrite](https://github.com/peak/s5cmd/blob/master/e2e/run_test.go#L266) the timeout. Maybe I can just increase it?

**Comment by igungor** [comment]
> > Thanks. Increased s5cmd test command timeout to 10 minutes but the issue is still there. I'm not sure what the problem is on Go 1.21 macos test.
> 
> Hi @igungor , if I am not wrong, I think this line [overwrite](https://github.com/peak/s5cmd/blob/master/e2e/run_test.go#L266) the timeout. Maybe I can just increase it or remove it?

It'd be great if you could remove the timeout line.

**Review by igungor — APPROVED** [review]

**Comment by igungor** [comment]
Thanks!

**Comment by vrdn-23** [comment]
@igungor @lizzzcai Now that the new version has been updated to docker, would it be possible to cut a new release so that we can pull it using the release tag?

**Comment by lizzzcai** [comment]
> @igungor @lizzzcai Now that the new version has been updated to docker, would it be possible to cut a new release so that we can pull it using the release tag?

Hi @vrdn-23 , I am expecting a new release as well. Probably need to check with the maintainer on the plan.

**Pull request merged**
